src/Transports.AspNetCore/WebSockets/IWebSocketConnection.cs

```diff
@@ -0,0 +1,50 @@
+using GraphQL.Transport;
+using Microsoft.AspNetCore.Http;
+
+namespace GraphQL.Server.Transports.AspNetCore.WebSockets;
+
+/// <summary>
+/// Represents a WebSocket connection, dispatching received messages over
+/// the connection to the specified <see cref="IOperationMessageProcessor"/>,
+/// and sending requested messages out the connection when requested.
+/// <br/><br/>
+/// <see cref="ExecuteAsync(IOperationMessageProcessor)"/> may be only called once.
+/// All members must be thread-safe.
+/// </summary>
+public interface IWebSocketConnection
+{
+    /// <summary>
+    /// Listens to incoming messages over the WebSocket connection,
+    /// dispatching the messages to the specified <paramref name="operationMessageProcessor"/>.
+    /// Returns or throws <see cref="OperationCanceledException"/> when the WebSocket connection is closed.
+    /// </summary>
+    Task ExecuteAsync(IOperationMessageProcessor operationMessageProcessor);
+
+    /// <summary>
+    /// Sends a message.
+    /// </summary>
+    Task SendMessageAsync(OperationMessage message);
+
+    /// <summary>
+    /// Closes the WebSocket connection.
+    /// </summary>
+    Task CloseConnectionAsync();
+
+    /// <summary>
+    /// Closes the WebSocket connection with the specified error information.
+    /// </summary>
+    Task CloseConnectionAsync(int eventId, string? description);
+
+    /// <summary>
+    /// Returns the last UTC time that a message was sent.
```

Comment:
From server to client? Why do you need it?

Reply:
To implement the idle timeout, it seemed to me to make sense to only send a keep-alive message after ## seconds of "no traffic".  But sent messages are posted to a queue, and there is no way for `IOperationMessageProcessor` to know when the last message was actually sent.  Multiple subscriptions may be active on one channel, not to mention queries or mutations being executed, and some `ExecutionResult` messages may be large, or connection quality may be poor -- any of those conditions may delay sending of messages from when it was posted.  So this property allows for a more accurate idle timeout implementation.